File: 1392-longest-happy-prefix/1392-longest-happy-prefix.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        i, j, n = 0, 1, len(s)
        lps = [0] * n
        while j < n:
            if s[i] == s[j]:
                lps[j] = i + 1
                i, j = i + 1, j + 1
            elif i == 0:
                lps[j] = 0
                j += 1
            else:
                i = lps[i - 1]
        prefixLen = lps[n - 1]
        return s[0 : prefixLen]

File: 1392-longest-happy-prefix/1392-longest-happy-prefix.cpp
class Solution {
public:
    string longestPrefix(string s) {
        // find out the longest prefix which is also a suffix
        int i = 0, j = 1, n = size(s);
        vector<int> lps(n);
        while(j < n) {
            // Case 1: both matching
            if(s[i] == s[j]) {
                lps[j] = i + 1;
                ++i;
                ++j;
             } 
            // Case 2: if the start is not even matching
             else if(i == 0) {
                lps[j] = 0;
                ++j;
             }
             // Case 3: the mid portion is not matching
             else {
                    // move i back
                    i = lps[i - 1];
            }
        }
        int prefixLen = lps.back();
        return s.substr(0, prefixLen);
    }
};